content/blog/2022/2022-05-27-one-stop-guide-mapstruct.md

```diff
@@ -0,0 +1,2116 @@
+    }
+
+    public void addManagerList(ManagerDTO managerDTO) {
+        if (managers == null) {
+            managers = new ArrayList<>();
+        }
+
+        managers.add(managerDTO);
+    }
+    
+    // other getters and setters
+}
+```
+
+Note that we have both the setter method, `setManagers`, and the adder method, `addManagerList` and we are responsible to initiate the collection for the adder. Then we have defined the default mapper the implementation looks something like the below:
+
+```java
+@Generated(
+    value = "org.mapstruct.ap.MappingProcessor"
+)
+public class UserMapperImpl implements UserMapper {
+
+    @Override
+    public PersonDTO map(Person person) {
+        if (person == null) {
+            return null;
+        }
+
+        PersonDTO personDTO = new PersonDTO();
+
+        personDTO.setManagerList(personMapper.map(person.getManagerList()));
+
+        return personDTO;
+    }
+}
+```
+
+As we can see, MapStruct uses setter method to set the `PersonDTO` instance. Since MapStruct uses the `ACCESSOR_ONLY` collection mapping strategy. But if we pass and attribute in `@Mapper` to use the `ADDER_PREFERRED` collection mapping strategy then it would look something like the below:
+
+```java
+@Mapper(collectionMappingStrategy = CollectionMappingStrategy.ADDER_PREFERRED,
+       uses = ManagerMapper.class)
+public interface PersonMapperAdderPreferred {
+    PersonDTO map(Person person);
+}
+```
+
+The generated implementation method would look something like the below:
+
+```java
+public class PersonMapperAdderPreferredImpl implements PersonMapperAdderPreferred {
+
+    private final ManagerMapper managerMapper = Mappers.getMapper( ManagerMapper.class );
+    
+    @Override
+    public PersonDTO map(Person person) {
+        if ( person == null ) {
+            return null;
+        }
+
+        PersonDTO personDTO = new PersonDTO();
+
+        if ( person.getManagerList() != null ) {
+            for ( Manager manager : person.getManagerList() ) {
+                personDTO.addManagerList( managerMapper.convert( manager ) );
+            }
+        }
+
+        return personDTO;
+    }
+}
+```
+
+In case the adder was not available, the setter would have been used.
+
+### Mapping Streams
+
+Mapping Streams are similar to mapping collections. The only difference is that the auto-generated implementation would return a `Stream` from a provided `Iterable`:
+
+```java
+@Mapper
+public interface CollectionMapper {
+    CollectionMapper INSTANCE = Mappers.getMapper(CollectionMapper.class);
+
+    Set<String> convertStream(Stream<Long> ids);
+
+    @Mapping(source = "degreeName", target = "degree")
+    @Mapping(source = "institute", target = "college")
+    @Mapping(source = "yearOfPassing", target = "passingYear")
+    EducationDTO convert(Education education);
+    List<EducationDTO> convert(Stream<Education> educationStream);
+}
+```
+
+The implementation methods would look something like below:
+
+```java
+@Generated(
```

Comment:
Not sure if we need the generated code here.

Reply:
Yes, this generated code is required as it describes about the conversion of Streams to List or Set and how it's handled.